### PR TITLE
DrivAerML: 788 batches/epoch + T_max=60 (proportionally scaled schedule)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

Your previous experiment (#3084) showed that 788 batches/epoch diverges because T_max=30 (epoch-based) effectively halves the cosine cycle with 2x batches, creating more frequent LR peak shocks. The fix: scale T_max proportionally. With 788 batches/epoch (2x more than champion's 394), T_max should be 2x longer (T_max=60) to maintain the same number of effective training steps per cosine cycle.

## Instructions

**Run: 788 batches + T_max=60 (proportionally scaled) + gc=1.0 for stability**
```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 \
  --cosine-t-max 60 --grad-clip 1.0 \
  --no-use-ema --enable-fourier \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 788 --max-eval-batches 200 \
  --wandb_group dm-batches-scaled-tmax
```

gc=1.0 is included for gradient stability at the higher batch count. Report best val_primary/surface_rel_l2_pct.

## Baseline
- **val_primary/surface_rel_l2_pct:** 3.997% at epoch 467
- **Your #3084 result:** 11.565% best (788 batches, T_max=30) — diverged at ep45

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT kakashi:

## Results

### 788 batches/epoch + T_max=60 + gc=1.0

| Metric | Baseline (394 batches, T_max=30) | This run (788 batches, T_max=60) | Previous #3084 (788 batches, T_max=30) |
|--------|---------|----------|----------|
| **val_primary/surface_rel_l2_pct** | **3.997%** (ep467) | **6.620%** (ep66) | **11.565%** (ep45) |

- **W&B run:** 7sa8l03q (kakashi/dm-788-tmax60)
- **Best epoch:** 66 (val 6.620%)
- **Divergence:** Epoch 72 — catastrophic gradient explosion (grad_norm 0.26 → 2.1 → 22.8 → 112 → 175 → 2314 → 2M → ∞)
- **Peak memory:** ~6 GB VRAM

**Command used:**
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 \
  --cosine-t-max 60 --grad-clip 1.0 \
  --no-use-ema --enable-fourier \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 788 --max-eval-batches 200 \
  --agent kakashi --wandb-name "kakashi/dm-788-tmax60" \
  --wandb-group dm-batches-scaled-tmax
```

## What happened

**T_max=60 significantly improved on T_max=30 at 788 batches** (6.62% vs 11.57%) — the proportional scaling hypothesis is confirmed in direction. The longer cosine cycle prevented the early divergence that T_max=30 caused. However, the model still diverged catastrophically at epoch 72 (start of the second cosine cycle), following the same pattern:

1. **Epochs 1-66:** Healthy training with steadily descending trough envelope (31.7% → 6.62%)
2. **Epoch 72:** Sudden gradient explosion — grad_clip_events jumped from 5 to 262, grad_norm from 0.28 to 2.1
3. **Epochs 73-78:** Cascading divergence (grad_norm 22 → 112 → 175 → 2314 → 2M → ∞)
4. **Epochs 78+:** Stuck at ~75.5%, irrecoverable

**The result does NOT beat the baseline** (6.62% vs 3.997%). The 394-batch baseline at T_max=30 trained for 467 epochs without diverging, giving it 7x more cosine cycles and far more opportunity for deep basin exploration. 788 batches sees 2x more data per epoch but the instability limits total training horizon.

**Key insight:** gc=1.0 was insufficient to prevent the second-cycle divergence at 788 batches. The gradient explosion at epoch 72 was sudden and catastrophic — not gradual. This suggests the instability is an abrupt phase transition rather than a slow accumulation, possibly triggered when the cosine LR peak coincides with a particularly high-gradient batch sequence.

## Suggested follow-ups

1. **Lower gc (0.5 or 0.3)** at 788 batches + T_max=60 — tighter clipping might prevent the epoch-72 blowup while preserving the faster convergence from more data per epoch
2. **EMA + 788 batches** — EMA could stabilize through the LR peaks (worked for TandemFoil stability)
3. **T_max=120** (further scaling) — even gentler LR cycling might push the divergence point out further
4. **Lower LR (3e-4 or 4e-4)** at 788 batches — the 788-batch regime may need a lower base LR to stay stable across multiple cycles

**Comment by tcapelle** [comment]
ADVISOR: The proportional scaling is directionally confirmed — 6.62% is much better than #3084's 11.57%. But gc=1.0 isn't tight enough: divergence at ep72 (second cosine cycle) shows the gradient peaks are still too large. Please retry with **--grad-clip 0.5** (matching the new DM champion gc level). Also add **--ema-decay 0.9995** — EMA is now mandatory for competitive DM results per PR #3072. Full config: 788 batches + T_max=60 + gc=0.5 + EMA=0.9995. This compound may finally make higher throughput work.

**Comment by tcapelle** [comment]
STUDENT kakashi: Running follow-up per advisor feedback — 788 batches + T_max=60 + gc=0.5 + EMA=0.9995. Fixed a rebase conflict (duplicate best-tracking code from PR #2974 shadowing `primary_metric_key` function). Training now running on GPU 0, W&B: kakashi/dm-788-tmax60-gc05-ema.

**Comment by tcapelle** [comment]
STUDENT kakashi:

## Results

### Run 2: 788 batches/epoch + T_max=60 + gc=0.5 + EMA=0.9995

| Metric | Baseline (394 batches, T_max=30, gc=0.5, EMA=0.9995) | This run (788 batches, T_max=60, gc=0.5, EMA=0.9995) | Run 1 (788 batches, T_max=60, gc=1.0, no-EMA) |
|--------|---------|----------|----------|
| **val_primary/surface_rel_l2_pct** | **3.833%** (ep511) | **4.661%** (ep141) | **6.620%** (ep66) |

- **W&B run:** jzg8os84 (kakashi/dm-788-tmax60-gc05-ema)
- **Best epoch:** 141 (val 4.661%)
- **Divergence:** Epoch 143 — gradient explosion (grad_norm 0.14 → 1.73 → 260 → 2316 → 2209, val 4.66% → 17.8% → 34.5% → 62.1%)
- **Peak memory:** ~6 GB VRAM

**Command used:**
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 \
  --cosine-t-max 60 --grad-clip 0.5 \
  --ema-decay 0.9995 --enable-fourier \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 788 --max-eval-batches 200 \
  --agent kakashi --wandb-name "kakashi/dm-788-tmax60-gc05-ema" \
  --wandb-group dm-batches-scaled-tmax
```

## What happened

**gc=0.5 + EMA=0.9995 dramatically extended stability at 788 batches** — from ep72 divergence (gc=1.0) to ep143, a 2x improvement in training horizon. Best val improved from 6.62% to 4.66% (30% relative improvement). The combination solved the second-cycle divergence that killed Run 1.

**Training phases:**
1. **Epochs 1-72:** Monotonic descent from 53% to 5.78%. 72 consecutive new bests. Grad norms steadily decreased from 7.2 to 0.27. gc=0.5 clip events dropped from 770/788 to 68/788 as gradients naturally settled.
2. **Epochs 73-141:** Oscillating descent in 4.7-7.1% range. Periodic transient spikes (ep106, 115-116, 123-125, 132-133) where grad norms briefly rose to 0.4-0.7 — all resolved within 2-3 epochs. Trough envelope continued descending: 5.08% (ep96) → 5.02% (ep113) → 4.87% (ep122) → 4.66% (ep141).
3. **Epoch 142-143:** Abrupt catastrophic divergence. Grad norm jumped 0.14 → 1.73 → 260 → 2316 in 3 epochs. Same exponential blowup pattern as Run 1 but delayed by 71 epochs. EMA=0.9995 could not absorb the explosion — val went 4.66% → 17.8% → 34.5% → 62.1%.

**The result does NOT beat the baseline** (4.66% vs 3.833%). The 394-batch baseline trained for 511 epochs without diverging (with the same gc=0.5+EMA config), giving it 3.6x more total epochs. The 788-batch regime sees 2x more data per epoch, but the instability limits total training to 141 epochs — resulting in 141×788 = 111k total steps vs the baseline's 511×394 = 201k steps (1.8x fewer total training steps).

**Key insight:** The 788-batch instability is not a cosine-cycle boundary effect (it appeared at ep143, mid-cycle). The periodic transient spikes (every ~10-15 epochs) suggest stochastic gradient perturbations that gc=0.5 can contain, but eventually one exceeds the containment threshold and cascades. The 394-batch baseline avoids this entirely because smaller batch windows produce less extreme individual gradient events.

**Bug fix note:** Fixed a rebase conflict from PR #2974 (best-checkpoint saving) where duplicate best-tracking code shadowed the `primary_metric_key` function, causing an `UnboundLocalError`. Removed the old tracking block and updated variable references.

## Suggested follow-ups

1. **gc=0.3 at 788 batches** — even tighter clipping might push the divergence point past ep200+ and allow the trough envelope to reach sub-4%
2. **Lower LR (3e-4 or 4e-4) at 788 batches** — the larger batch window may need a lower base LR to avoid the stochastic gradient perturbations
3. **T_max=120 at 788 batches** — gentler LR cycling to reduce the cosine peak stress
4. **Abandon 788-batch line** — the 394-batch config is fundamentally more stable and already achieves 3.833%. The 2x data/epoch advantage of 788 batches doesn't compensate for the training horizon limitation. The diminishing returns suggest the stability ceiling is intrinsic to the batch count, not fixable by hyperparameter tuning.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #3138 because ADVISOR: 788-batch regime is a dead end — three attempts all diverge catastrophically within 2 cosine cycles. Run 1 (#3084, T_max=30): diverged early. Run 2 (gc=1.0, T_max=60): 6.620% ep66, diverged ep72. Run 3 (gc=0.5+EMA, T_max=60): 4.661% ep141, diverged ep143. The champion platform extended stability from ep72→ep143 and improved from 6.62%→4.66%, but divergence is structural — 788 batches accumulate only 111k total steps before dying vs baseline's 201k. Your analysis that the instability is intrinsic to the 788-batch regime is correct. GPU time better spent elsewhere..
